test/UnitTest/Microsoft.AspNetCore.OData.Test/Abstraction/ODataActionSelectorTestHelper.cs

```diff
@@ -0,0 +1,155 @@
+            var modelMetadataProvider = serviceProvider.GetRequiredService<IModelMetadataProvider>();
+            actionSelector = new ODataActionSelector(
+                actionsProvider,
+                actionConstraintsProvider,
+                loggerFactory,
+                modelBinderFactory,
+                modelMetadataProvider);
+        }
+#else
+        {
+            var innerActionSelectorMock = new Mock<IActionSelector>();
+            routeBuilder = RoutingConfigurationFactory.Create();
+            var serviceProvider = routeBuilder.ServiceProvider;
+            var modelBinderFactory = (IModelBinderFactory)serviceProvider.GetService(typeof(IModelBinderFactory));
+            var modelMetadataProvider = (IModelMetadataProvider)serviceProvider.GetService(typeof(IModelMetadataProvider));
+            actionSelector = new ODataActionSelector(innerActionSelectorMock.Object, modelBinderFactory, modelMetadataProvider);
+            actionDescriptors = ControllerDescriptorFactory.Create(routeBuilder, controllerType.Name, controllerType)
+                as IReadOnlyList<ControllerActionDescriptor>;
+            actionDescriptors = actionDescriptors.Where(a => a.ActionName == actionName).ToList();
+        }
+#endif
+
+        /// <summary>
+        /// Create route context with the specified with the specified request data
+        /// </summary>
+        /// <param name="routeBuilder"></param>
+        /// <param name="actionName">Name of the action ebging routed to</param>
+        /// <param name="routeDataValues">Key-value pairs to add to the route data</param>
+        /// <param name="method">HTTP request method</param>
+        /// <param name="bodyContent">Request body content</param>
+        /// <returns></returns>
+        public static RouteContext SetupRouteContext(IRouteBuilder routeBuilder, string actionName,
+            Dictionary<string, object> routeDataValues,string method, string bodyContent)
+        {
+            var request = RequestFactory.Create(routeBuilder);
+            var routeContext = new RouteContext(request.HttpContext);
+            var routeData = routeContext.RouteData;
+            var routingConventionsStore = routeContext.HttpContext.ODataFeature().RoutingConventionsStore;
+            var odataPath = new ODataPath();
+            routeContext.HttpContext.ODataFeature().Path = odataPath;
+            
+            routeData.Values[ODataRouteConstants.ODataPath] = odataPath;
+            routeData.Values[ODataRouteConstants.Action] = actionName;
+            int keyCount = 0;
+            foreach (var keyValuePair in routeDataValues)
+            {
+                routeData.Values[keyValuePair.Key] = keyValuePair.Value;
+                keyCount++;
+            }
+
+            routingConventionsStore[ODataRouteConstants.KeyCount] = keyCount;
+
+            request.Method = method;
+
+            if (bodyContent != null)
+            {
+                var stream = new MemoryStream();
```

Comment:
@habbes Easier way to achieve this: `var stream = new MemoryStream(Encoding.UTF8.GetBytes(bodyContent));`